google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java

```diff
@@ -250,4 +269,70 @@ ParametersInfo convertPositionalParametersToNamedParametersInternal(char paramCh
     }
     return new ParametersInfo(paramIndex - 1, named.toString());
   }
+
+  private boolean isReturning(String sql, int index) {
+    return Pattern.compile("[\\s)]THEN(\\s+)RETURN[\\s(][\\s\\S]*", Pattern.CASE_INSENSITIVE)
+        .matcher(sql.substring(index))
+        .matches();
```

Comment:
Also here: I think this could be made more efficient by at least removing the `compile` statement out of this method, but I also think that you don't need a regex at all. You can just do a simple string check if you first replace all double spaces with single spaces, and make the SQL string all lower or upper case.

Reply:
Added answer to this in https://github.com/googleapis/java-spanner/pull/1978#discussion_r952676785.